[TorchElastic] Option for sharing TCPStore created by rdzv handlers

kurman opened the pull request:
Summary:

1. Define explicit `use_agent_store` on rdzv handlers. Handlers that set is true can share the store. 
2. Instead of agent coordinating master_add/master_port values, the logic is now encapsulated by a *rdzv_handler* where `RendezvousInfo` will have `RendezvousStoreInfo` object that handlers must return. 
    - Depending on the implementation they can either:
         - point to existing store (and expected to `use_agent_store` as true - point 1). Client code will rely on `TORCHELASTIC_USE_AGENT_STORE` env variable to know if the store is shared.
         - build args that `torch.distributed.init_process_group` can bootstrap by creating new store. 

Additional points:

- When TCPStore is shared, it should be wrapped in PrefixStore to qualify/scope namespace for other usecases.
- `next_rendezvous` signature changed to return instance of `RendezvousInfo` instead of a (store, rank, world_size) tuple for extensibility purposes.


Why:
- Reduce moving parts
   - easier to swap implementation
   - improve tractability
   - addressing perf/debug-ability will benefit all usecases
   - 
Test Plan: CI

Differential Revision: D57055235


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

d4l3k commented on torch/distributed/elastic/agent/server/api.py:
```diff
@@ -557,12 +557,20 @@ def _rendezvous(self, worker_group: WorkerGroup) -> None:
+            if (
+                os.getenv("TORCH_SHARE_RDZV_TCP_STORE", "0") == "1"
+                and spec.rdzv_handler.static
+                and store.check(["MASTER_ADDR", "MASTER_PORT"])
```
I'm not a fan of using the store to pass data between the rendezvous and agent. Since we're adding new methods/properties to the rendezvous class can we add one to get the master addr + port instead? That could also help clean up the existing static rendezvous logic (i.e. nuke spec.master_addr/master_port)

d4l3k commented on torch/distributed/elastic/agent/server/api.py:
```diff
@@ -557,12 +557,20 @@ def _rendezvous(self, worker_group: WorkerGroup) -> None:
-            )
+            # pyre-ignore [16]: Undefined attribute: torch._C._distributed_c10d.Store` has no attribute `check
+            if (
+                os.getenv("TORCH_SHARE_RDZV_TCP_STORE", "0") == "1"
```
I'd prefer if this was an implementation detail within the rendezvous handler -- i.e. `if rdzv_handler.get_master_addr()`.

kurman commented:
Some of the actions are failing due module visibility issue check - addressing it.

d4l3k commented on torch/distributed/elastic/rendezvous/api.py:
```diff
@@ -35,6 +38,61 @@ class RendezvousGracefulExitError(RendezvousError):
+        self._store = store
+        self._rank = rank
+        self._world_size = world_size
+        self._bootstrap_store_info = bootstrap_store_info or RendezvousStoreInfo.build(rank=self._rank, store=self._store)
```
I don't like that we're hiding store operations in the constructor of this -- I think this should be a pure data class. If the rendezvous wants to do this operation it should explicitly call it and return it

d4l3k commented on torch/distributed/elastic/rendezvous/api.py:
```diff
@@ -48,19 +106,26 @@ class RendezvousHandler(ABC):
         """Return the name of the rendezvous backend."""
 
+    @property
+    def use_agent_store(self) -> bool:
```
thoughts on using this vs making `_bootstrap_store_info` optional?

d4l3k commented on torch/distributed/elastic/agent/server/local_elastic_agent.py:
```diff
@@ -273,10 +273,9 @@ def _start_workers(self, worker_group: WorkerGroup) -> Dict[int, Any]:
         restart_count = spec.max_restarts - self._remaining_restarts
 
-        use_agent_store = spec.rdzv_handler.get_backend() == "static"
+        use_agent_store: bool = spec.rdzv_handler.use_agent_store
```
we don't actually use this anywhere?

d4l3k replied:
oh -- nvm, I see where it's specified below

d4l3k commented on torch/distributed/elastic/agent/server/api.py:
```diff
@@ -546,7 +486,16 @@ def _rendezvous(self, worker_group: WorkerGroup) -> None:
+
+        # master_addr/master_port could be explicitly overriden
+        # TODO: BC - specific to static rdzv and can be simplifed further
+        master_addr = spec.master_addr or rdzv_info.bootstrap_store_info.master_addr
```
should we be checking use_agent_store ?

kurman replied:
No, at this time it just needs master_addr and master_port regardless if store is shared or not

d4l3k reviewed (changes requested):
looking much better!

wconstab commented on torch/distributed/elastic/rendezvous/api.py:
```diff
@@ -35,6 +38,61 @@ class RendezvousGracefulExitError(RendezvousError):
+    master_port: int
+
+    @staticmethod
+    def build(rank: int, store: Store) -> "RendezvousStoreInfo":
```
I'm confused about the order of operations.

I thought the build() path was used when we did not have a store to share from the elastic later.

Actually I guess it must be the following- we always have a tcpstore at the elastic layer, so we can use it to share information. But we call build() when we don't want to share that store with the trainer, and instead we just use the elastic's store for exchanging port info so the trainer can create its own store?

kurman replied:
Yes. Build here is a factory method of the `RendezvousStoreInfo` that will define `master_addr/master_port` env variables that trainer code can use to start a new store server.

kurman committed:
[TorchElastic] Option for sharing TCPStore created by rdzv handlers (#125743)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/125743

1. Define explicit `use_agent_store` on rdzv handlers. Handlers that set is true can share the store.
2. Instead of agent coordinating master_add/master_port values, the logic is now encapsulated by a *rdzv_handler* where `RendezvousInfo` will have `RendezvousStoreInfo` object that handlers must return.
    - Depending on the implementation they can either:
         - point to existing store (and expected to `use_agent_store` as true - #1). Client code will rely on `TORCHELASTIC_USE_AGENT_STORE` env variable to know if the store is shared.
         - build args that `torch.distributed.init_process_group` can bootstrap by creating new store.

Additional points:

- When TCPStore is shared, it should be wrapped in PrefixStore to qualify/scope namespace for other usecases.
- `next_rendezvous` signature changed to return instance of `RendezvousInfo` instead of a (store, rank, world_size) tuple for extensibility purposes.

Why:
- Reduce moving parts
   - easier to swap implementation
   - improve tractability
   - addressing perf/debug-ability will benefit all usecases

Test Plan: CI

Differential Revision: D57055235

d4l3k reviewed (approved):
LGTM

d4l3k commented on test/distributed/elastic/agent/server/test/api_test.py:
```diff
@@ -298,7 +302,8 @@ def _get_record_metrics_test_calls(
     def test_rendezvous(self):
-        spec = self._get_worker_spec(max_restarts=1)
+        hostname = _get_fq_hostname()
+        spec = self._get_worker_spec(max_restarts=1, local_addr=hostname)
```
I see this test is changing? Are we changing behavior here around requiring the fqdn being passed in?

kurman replied:
Collectives operation is happening in the handler now so we need a address that is resolvable.
Prior to this changes tests would just query store for master_addr/master_port on the store.

Yejing-Lai commented on torch/distributed/elastic/agent/server/api.py:
```diff
@@ -356,37 +357,6 @@ def is_failed(self) -> bool:
         return self.state == WorkerState.FAILED
 
 
-def _get_socket_with_port() -> socket.socket:
```
Hi @kurman. After delete '_get_socket_with_port' api. Can I know what is the equivalent API/method?

kurman replied:
@Yejing-Lai sorry missed your note. Would `from torch.distributed.elastic.utils.distributed import get_free_port` work? rendezvous/api.py uses the function now.
